Add rate limiting for check_username_availability

nospame opened the pull request:
## Product Description
Adds rate limiting when checking username availability prior to signup. Invisible under normal circumstances.

## Technical Summary
[SAAS-16420
](https://dimagi.atlassian.net/browse/SAAS-16420)
Based on [rate_limit_two_factor_setup](https://github.com/dimagi/commcare-hq/blob/8802076f90e2b0682dc27c312977179822740582/corehq/apps/hqwebapp/two_factor_gateways.py#L103), this adds a rate limiter for the `check_username_availability` endpoint that is based on both IP (more lax, since public IPs can sometimes be shared) and session ID (more strict). 

Specific rates here were chosen because they seemed "about right" for someone possibly typing and retyping an email address several times during signup, with the fastest per-session rate matching the [frontend rate limit](https://github.com/dimagi/commcare-hq/blob/8802076f90e2b0682dc27c312977179822740582/corehq/apps/registration/static/registration/js/new_user.ko.js#L83) on the signup form (a 400ms delay). Open to other thoughts here as well.

## Safety Assurance

### Safety story
Tested locally to see that IP and session ID get stored and checked against as expected for rate limiting, and that rate limits do seem to apply correctly when requests are quick enough.

### Automated test coverage
No automated tests here.

### QA Plan
Not currently planning QA.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Add rate limiting for check_username_availability

mjriley commented on corehq/apps/registration/utils.py:
```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
         raise
+
+
+@run_only_when(not settings.ENTERPRISE_MODE and not settings.UNIT_TESTING)
```
Should we enforce this for non-enterprise environments as well?

nospame replied:
https://github.com/dimagi/commcare-hq/pull/35688/commits/b48eb3154e26d569573d39cdd49cf3beb870de96

biyeun commented on corehq/apps/registration/utils.py:
```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
+        if request and request.session:
+            return request.session.session_key, get_ip(request)
+        else:
+            return None
```
should return `None, None`

nospame replied:
https://github.com/dimagi/commcare-hq/pull/35688/commits/b48eb3154e26d569573d39cdd49cf3beb870de96

mjriley commented on corehq/apps/registration/utils.py:
```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
+        if request and request.session:
+            return request.session.session_key, get_ip(request)
+        else:
+            return None
```
needs to return `None, None`

mjriley commented on corehq/apps/registration/utils.py:
```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
+
+    session_id, ip_address = get_session_and_ip()
+
+    if ip_address and session_id:
```
Given that these are bad-faith actors, is it possible to continually create new session ids? Is this a concern?

nospame replied:
Not as far as I've found. If given a bad session ID, the request is rejected on the backend and the session ID cookie gets removed.

mjriley commented on corehq/apps/registration/utils.py:
```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
+
+    Requests without both an IP and session ID are rejected.
+    """
+    _status_session_rate_limited = 'session_rate_limited'
+    _status_ip_rate_limited = 'ip_rate_limited'
+    _status_bad_request = 'bad_request'
+    _status_accepted = 'accepted'
```
I think our standard for constants is all-caps. Can also move these to a const module.

nospame replied:
Didn't move to a const module, but did move the rate limiter to its own file, as we seem to do in other apps (ota, motech, receiverwrapper)
https://github.com/dimagi/commcare-hq/pull/35688/commits/7f9d3289ddbf9854fbbf2b06b2a0a9d80989cc8d

nospame committed:
Move check_username_availability rate limiter to rate_limiter.py

This pattern is used in other apps with rate limiters

nospame committed:
Break out functions from rate_limit_check_username_availability and use constants

nospame committed:
Address feedback:

- run rate limiter in Enterprise environments
- return None, None fallback from _get_session_and_ip
- return isValid: False if check_username_availability is rate limited

nospame commented on corehq/apps/registration/utils.py:
```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
+        else:
+            return None
+
+    def _check_for_exceeded_rate_limits(session, ip):
```
https://github.com/dimagi/commcare-hq/pull/35688/commits/7f9d3289ddbf9854fbbf2b06b2a0a9d80989cc8d

nospame commented on corehq/apps/registration/utils.py:
```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
+    _status_bad_request = 'bad_request'
+    _status_accepted = 'accepted'
+
+    def get_session_and_ip():
```
https://github.com/dimagi/commcare-hq/pull/35688/commits/7f9d3289ddbf9854fbbf2b06b2a0a9d80989cc8d

mjriley reviewed (approved):
LGTM
